compiler/src/dotty/tools/dotc/core/TypeUtils.scala

```diff
@@ -128,6 +128,7 @@ class TypeUtils:
         case None => throw new AssertionError("not a tuple")
 
     def namedTupleElementTypesUpTo(bound: Int, normalize: Boolean = true)(using Context): List[(TermName, Type)] =
+      if bound < 0 then Nil else
```

Comment:
I think that's a bit too coarse. You recurse 2^31 times in the normal case. It's a bit tricky to limit recursion in the face of illegal recursive aliases, because we do the `namedTupleElementTypesUpTo` before Typer in desugaring. There we have not yet checked and ruled out such illegal cycles. 

One idea would be to leave namedTupleElementTypesUpTo as it is for desugaring, but have another version that follows superTypes of TypeProxies in Typer and Completions. Or have another Boolean flag that says whether we should follow proxies and &/I.

Reply:
> One idea would be to leave namedTupleElementTypesUpTo as it is for desugaring, but have another version that follows superTypes of TypeProxies in Typer and Completions. 

It's not clear to me why desugaring should be treated specially, in https://github.com/scala/scala3/blob/0677702aa4b73dea8740a66296423747fd9bf4ee/compiler/src/dotty/tools/dotc/ast/Desugar.scala#L1747-L1748 immediately after calling `pt.namedTupleElementTypes` we call `pt.classSymbol` and `classSymbol` also recurses on TypeProxy: https://github.com/scala/scala3/blob/0677702aa4b73dea8740a66296423747fd9bf4ee/compiler/src/dotty/tools/dotc/core/Types.scala#L580-L581

Reply:
But `selNames.isEmpty` will be false